py_snappy/main.py

```diff
@@ -0,0 +1,342 @@
+    bytes n is <= 0 meaning:
+    n == 0: buf too small
+    n  < 0: value larger than 64 bits (overflow)
+            and -n is the number of bytes read"""
+    x, s = 0, 0
+    for idx, b in enumerate(buf):
+        if b < 0x80:
+            if idx > 9 or (idx == 9 and b > 1):
+                return 0, -1 * (idx + 1)  # overflow
+            return x | uint64(b) << s, idx + 1
+        x |= uint64(b & 0x7F) << s
+        s += 7
+    return 0, 0
+
+
+TReturn = TypeVar("TReturn")
+
+
+def bytes_gen(fn: Callable[..., Iterable[int]]) -> Callable[..., bytes]:
+    @functools.wraps(fn)
+    def inner(*args: Any, **kwargs: Any) -> bytes:
+        return bytes(fn(*args, **kwargs))
+
+    return inner
+
+
+def tuple_gen(
+    fn: Callable[..., Iterable[TReturn]]
+) -> Callable[..., Tuple[TReturn, ...]]:
+    @functools.wraps(fn)
+    def inner(*args: Any, **kwargs: Any) -> Iterable[TReturn]:
+        return tuple(fn(*args, **kwargs))
+
+    return inner
+
+
+@bytes_gen
+def putuvarint(x: int) -> Iterable[int]:
+    """
+    putuvarint encodes a uint64.
+    """
+    while x >= 0x80:
+        yield uint8(x) | 0x80
+        x >>= 7
+    yield x
+
+
+def extract_meta(src: bytes) -> Tuple[int, int]:
+    """
+    Return a 2-tuple:
+
+    - the length of the decoded block
+    - the number of bytes that the length header occupied.
+  """
+    v, n = uvarint(src)
+    if n <= 0 or v > 0xFFFFFFFF:
+        raise CorruptError
+    if v > 0x7FFFFFFF:
+        raise TooLargeError
+    return v, n
+
+
+def decompress(buf: bytes) -> bytes:
+    """
+    decompress returns the decompressed form of buf.
+    """
+    block_length, s = extract_meta(buf)
+    src = tuple(c for c in buf)
+    src_len = len(src)
+    dst = [0] * block_length
+    d, offset, length = 0, 0, 0
+
+    while s < src_len:
+        b = src[s] & 0x03
+        if b == TAG_LITERAL:
+            x = src[s] >> 2
+
+            if x < 60:
+                s += 1
+            elif x == 60:
+                s += 2
+                if s > src_len:
+                    raise CorruptError
+                x = src[s - 1]
+            elif x == 61:
+                s += 3
+                if s > src_len:
+                    raise CorruptError
+                x = src[s - 2] | (src[s - 1] << 8)
+            elif x == 62:
+                s += 4
+                if s > src_len:
+                    raise CorruptError
+                x = src[s - 3] | (src[s - 2] << 8) | (src[s - 1] << 16)
+            elif x == 63:
+                s += 5
+                if s > src_len:
+                    raise CorruptError
+
+                x = (
+                    src[s - 4]
+                    | (src[s - 3] << 8)  # noqa: W503
+                    | (src[s - 2] << 16)  # noqa: W503
+                    | (src[s - 1] << 24)  # noqa: W503
+                )
+
+            length = x + 1
+
+            if length <= 0:
+                raise BaseSnappyError("Unsupported literal length")
```

Comment:
Something feels weird about raising the base error. Should this be a `ValidationError` maybe? Or perhaps there should be a custom `SnappyValidationError`.

Reply:
This is a leftover from the original implementation, will clean up all the exceptions to no longer have any places that raise `BaseSnappyError` since it's not really intended to be used directly (and it isn't a contextually helpful message).